### DISCUSSION TITLE
Why use   [].slice.call()

### DISCUSSION THREAD

**Discussion by ZhangChengLin** [twbs/bootstrap#31152] [General] [answered]
https://github.com/twbs/bootstrap/blob/v5.0.0-alpha1/site/content/docs/5.0/components/navs.md#L529

https://github.com/twbs/bootstrap/blob/v5.0.0-alpha1/site/content/docs/5.0/components/list-group.md#L341

```
var triggerTabList = [].slice.call(document.querySelectorAll('#myTab a'))
```

---

Why use this way of writing, just use the following \"document.querySelectorAll('#myTab a')\"?
What is the difference between them?

**Comment by ZhangChengLin** [comment]
![{221EC107-7AEE-41A4-8D68-0020621FABCF}_20200623174435](https://user-images.githubusercontent.com/30773389/85388905-46b29380-b579-11ea-9156-bb76bd7cc172.jpg)

**Comment by rohit2sharma95** [comment, marked as answer]
`Array.prototype.slice.call` is one way to convert the NodeList into a real array so that you can use the methods from `Array.prototype`. NodeList acts like an array, but doesn’t have the methods from `Array.prototype`.

**Reply by ZhangChengLin** [reply]
> `Array.prototype.slice.call` is one way to convert the NodeList into a real array so that you can use the methods from `Array.prototype`. NodeList acts like an array, but doesn’t have the methods from `Array.prototype`.


querySelectorAll returns an array.

**Reply by rohit2sharma95** [reply]
Nope @zhangchenglin! `querySelectorAll()` does not return an Array instead it returns a NodeList. And NodeList is not Array.
As per [MDN Docs](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll):

> The Document method `querySelectorAll()` returns a static (not live) NodeList representing a list of the document's elements that match the specified group of selectors.

You can use the `Array.from` as well, instead of `Array.prototype.slice.call`. These are the way of converting an array-like object to an array. You can not use the methods on Nodelist directly that are in the prototype of the Array. For example, you can not use the `map()` method on the NodeList directly until you convert it to an array. `forEach` is built-in in the NodeList though. And if you don't want to convert the NodeList to an array, you can use the classic for loop easily :smile:

**Reply by ZhangChengLin** [reply]
> Nope @zhangchenglin! `querySelectorAll()` does not return an Array instead it returns a NodeList. And NodeList is not Array.
> As per [MDN Docs](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll):
> 
> > The Document method `querySelectorAll()` returns a static (not live) NodeList representing a list of the document's elements that match the specified group of selectors.
> 
> You can use the `Array.from` as well, instead of `Array.prototype.slice.call`. These are the way of converting an array-like object to an array. You can not use the methods on Nodelist directly that are in the prototype of the Array. For example, you can not use the `map()` method on the NodeList directly until you convert it to an array. `forEach` is built-in in the NodeList though. And if you don't want to convert the NodeList to an array, you can use the classic for loop easily 😄


https://codepen.io/zhangchenglin/pen/eYJWEWV?editors=1010

Doesn't affect use here? Is there any performance loss?
I am wondering why the for loop is not used directly.

![{A6FA4D0F-B033-4F12-B5E0-D6FB4EB9104A}_20200623202303](https://user-images.githubusercontent.com/30773389/85403275-a87df800-b58f-11ea-9169-d829b6bafa4c.jpg)


![{105E709C-8157-44A2-8524-685564A55D68}_20200623202742](https://user-images.githubusercontent.com/30773389/85403555-26420380-b590-11ea-82b8-cefdba1d47d0.jpg)

![{34D08992-02F1-4052-BA9D-0A185C031A3D}_20200623202808](https://user-images.githubusercontent.com/30773389/85403547-23471300-b590-11ea-855b-3b8d56e587ab.jpg)

**Reply by ZhangChengLin** [reply]
https://developer.mozilla.org/zh-CN/docs/Web/API/NodeList

However, some older browsers have not implemented NodeList.forEach() nor Array.from(). This can be circumvented by using Array.prototype.forEach() — see this document's Example.
